Interruptible Implementation (and bonus bugfixes)

djspiewak opened the pull request:
Implemented `interruptible`, which is cool. Also fixed some bugs, which is even cooler. Also unleashed a firestorm of confusion about semantics, which became #987 

Depends on #984

djspiewak committed:
Entirely ignore Sync.Type on javascript

djspiewak committed:
Implementation of IO.interruptible

djspiewak commented:
I know what's happening here. It's really tricky and loosely related to #987. The problem is the *continuation* of an `Async` within an `Async` finalizer. Like, for example, the one within `interruptible`'s implementation. When an `Async` finalizer is itself an `Async` and the outer `Async` is canceled while suspended, the `Async` finalizer will run immediately. If that finalizer is suspended after registration fully completes, it will resume and then *immediately check its cancelation flag… which will be `true`, because its fiber was canceled. We have no particular flag for "finalizing", which I think we probably need.

This, btw, I think was the reason that I call `done` early: so that it can be checked in `Async`'s `cur.k` implementation. With that said, we don't *need* to check `outcome` there since we pass through the `suspended` write barrier (when we `resume()`), so we can just rely on `canceled`. So this is the relation to the other ticket.

But we really do need a `finalizing` state, since it changes the meaning of so many things, and inferring it is tricky.

djspiewak commented:
As an aside, I did some digging, and part of the reason we set `outcome` early is so that we can check it in `loop` and `asyncContinue`. When an `Async` finishes after cancelation starts (but before it completes), we don't want to fork the runloop. This is easily avoided by `suspended`. However, when cancelation has *already* completed… and then the `Async` completes… we need to prevent the runloop from being resurrected. This is not so easily accomplished, and even setting `masks = initMask` and allowing `canceled` to remain `true` doesn't seem to be enough. I think being more explicit about this particular scenario is what's needed.

djspiewak committed:
Merge remote-tracking branch 'raas/backpressured-cancels' into feature/interruptible

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak commented:
Depends on #998

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak committed:
Fixed race condition where interrupt happens before callback registration

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IOFiber.scala:
```diff
@@ -330,28 +333,99 @@ private[effect] final class IOFiber[A](
+                        nextCb(Right(cur.thunk()))
+                      } catch {
+                        case _: InterruptedException =>
+                          println("caught interrupt")
```
stray println

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IOFiber.scala:
```diff
@@ -330,28 +333,99 @@ private[effect] final class IOFiber[A](
+                  }
 
-                runLoop(next, nextIteration)
+                  isSelf <- IO(target eq Thread.currentThread())
```
So the only way this could happen on the JVM is if we're using the same thread pool for both `compute` and `blocking`, correct? And if that's the case, we know for a fact that the blocking operation has completed, because the thread would have had to complete the operation above in order to reach this point, so there's nothing to interrupt anymore

djspiewak replied:
Yep! This is like a random emergency check to make sure we don't interrupt ourself, since that's one situation in which things would dramatically fall apart, and also a situation which is relatively easy to check.

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IOFiber.scala:
```diff
@@ -330,28 +333,99 @@ private[effect] final class IOFiber[A](
+                  } else {
+                    Some {
+                      IO async { finCb =>
+                        val trigger = IO {
```
What happens if the blocking thread completes its operation, and it picks up another task, but before the finalizer is de-registered, we `cancel`, which runs the finalizer? We might be interrupting or killing a `Thread` that's running someone else's work now.

If we use `done` to communicate that the blocking operation is or isn't done, then I think we still have a race condition, where the finalizer observes that `done` is false, so it decides to `Thread.interrupt()`, but before it actually starts to interrupt, the thread finishes the blocking operation and picks up another task.

djspiewak replied:
This is a very *very* good point. I need to think about this, but you're 100% right.

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IOFiber.scala:
```diff
@@ -330,28 +333,99 @@ private[effect] final class IOFiber[A](
+                          } else {
+                            val cb0 = cb.get()
+                            if (cb0 != null) {
+                              cb0()
```
If I'm understanding correctly, this is how we backpressure cancellers on interruption (particularly when we `Thread.interrupt` only once)?

djspiewak replied:
Yep!

djspiewak committed:
Removed stray println

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak committed:
Fixed race condition between interruption and cancelation

djspiewak commented:
@RaasAhsan Ready for another look! I think I solved the race condition.

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IOFiber.scala:
```diff
@@ -339,8 +343,118 @@ private[effect] final class IOFiber[A](
+                          // if done is false, and we can't get the semaphore, it means
+                          // that the action hasn't *yet* started, so we busy-wait for it
+                          var break = false
+                          while (break && !done.get()) {
```
Is this while loop dead code? `false && x == false`

djspiewak replied:
…how did the tests not catch this

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IOFiber.scala:
```diff
@@ -339,8 +343,118 @@ private[effect] final class IOFiber[A](
+
+                      try {
+                        canInterrupt.release(1)
+                        nextCb(Right(cur.thunk()))
```
Should we only run `cur.thunk()` in the interruptible region? Just so that nothing in `nextCb` is interrupted (I think the only possibility is a blocking `put` on the underlying task queue)

djspiewak replied:
Good call!

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IOFiber.scala:
```diff
@@ -339,8 +343,118 @@ private[effect] final class IOFiber[A](
+                  done <- IO(new AtomicBoolean(false))
+                  cb <- IO(new AtomicReference[() => Unit](null))
+
+                  canInterrupt <- IO(new juc.Semaphore(1))
```
Do we want to start the semaphore at 0 permits? The execution I'm thinking about is, what if we interrupt before the blocking operation is run? `tryAcquire` would be able to acquire a permit

djspiewak replied:
Excellent point

djspiewak committed:
Moved interruptible(...) implementation to platform-specific code

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak committed:
Fixed early cancelation of interruption

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak committed:
Scalafmt

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak commented:
@RaasAhsan Ready for another look. I'm pretty sure it actually works correctly now, though it does lean somewhat on an implementation detail of `IO`, which is that we check for cancelation at the *beginning* of evaluating a node, rather than at the end, meaning that the finalizer always gets registered properly.

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

RaasAhsan commented:
I will take a look at this as soon as I can, been a little bit busy

djspiewak commented:
> I will take a look at this as soon as I can, been a little bit busy

Absolutely no worries!

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

djspiewak committed:
Merge branch 'series/3.x' into feature/interruptible

RaasAhsan commented on core/jvm/src/main/scala/cats/effect/IOFiberPlatform.scala:
```diff
@@ -0,0 +1,141 @@
+                if (canInterrupt.tryAcquire()) {
+                  try {
+                    target.interrupt()
+                    break = false
```
Should we move this to the `finally` block, or before the `interrupt` call? `Thread.interrupt` can supposedly throw a `SecurityException` so this could in fact interrupt more than once.

RaasAhsan commented on core/jvm/src/main/scala/cats/effect/IOFiberPlatform.scala:
```diff
@@ -0,0 +1,141 @@
+          IO async { finCb =>
+            val trigger = IO {
+              if (!many) {
+                cb.set(() => finCb(Right(())))
```
I don't feel too strongly about it, but could avoid 2 allocations here if we just pass in `finCb` and allocate `Right(())` once similar to whats done in `IOFiber`

RaasAhsan commented on core/jvm/src/main/scala/cats/effect/IOFiberPlatform.scala:
```diff
@@ -0,0 +1,141 @@
+
+    IO.async[Any] { nextCb =>
+      for {
+        done <- IO(new AtomicBoolean(false))
```
Is it worth allocating `done`, `cb` and `canInterrupt` once as instance variables?

djspiewak replied:
Hmm, that would certainly speed up `interruptible`, but at the cost of slowing down `start` and making fibers somewhat more heavyweight, even when `interruptible` isn't used. Is it worth it? `interruptible` doesn't seem like a particularly commonly-needed operation.

RaasAhsan commented on core/jvm/src/main/scala/cats/effect/IOFiberPlatform.scala:
```diff
@@ -0,0 +1,141 @@
+            }
+
+            val repeat = if (many) {
+              IO blocking {
```
Should we be busy spinning on a blocking pool?

djspiewak replied:
Hmm. Good point. This should be on the compute pool.

RaasAhsan commented on core/jvm/src/main/scala/cats/effect/IOFiberPlatform.scala:
```diff
@@ -0,0 +1,141 @@
+          IO async { finCb =>
+            val trigger = IO {
+              if (!many) {
+                cb.set(() => finCb(Right(())))
```
Okay, so what happens when the blocking operation above completes, but the the finalizer is not yet popped, and some actor tries to cancel, interrupting only once? The callback is set in `cb` but it would never be invoked, so the canceller is (asynchronously) blocked forever.

djspiewak replied:
Nice catch! Fixed.

djspiewak committed:
Resolve race condition where canceler could block forever

djspiewak committed:
The interrupt busy-wait should be on the compute pool, since it doesn't actually block

djspiewak committed:
Ensure we break even if interrupt() throws

djspiewak committed:
Removed some allocation from interruptible

djspiewak committed:
Scalafmt

djspiewak commented:
@RaasAhsan Addressed review comments aside from the instance allocation of `done` etc, which merits further discussion. I lean *slightly* 👎 on that one.

RaasAhsan commented on core/jvm/src/main/scala/cats/effect/IOFiberPlatform.scala:
```diff
@@ -0,0 +1,149 @@
+              }
+            } else {
+              IO {
+                if (done.get() && cb.get() != null) {
```
Apologies to do this to you again... :) But what if the blocking operation above succeeds and `done` is not yet set to `true`? This can happen even with the `Semaphore` protection because `cur.thunk()` can reset the interrupt status in response to the first interrupt, so `canInterrupt.acquire()` doesn't throw an `InterruptedException`.

I haven't thought too much about it, but maybe we can fix this if we `getAndSet` and invoke  `cb` in the `finally` block, before setting `done`? I'll wait to see what you come up with :)

djspiewak replied:
I *believe* this case is actually handled! Note that the blocking operation is wrapped in a `try`/`finally` in which `done` is set to `true`. If the operation succeeds without being interrupted, then `cb` will be non-`null`, which will trip this conditional.

RaasAhsan replied:
Here is the ordering that I'm concerned about:
1. At line 64, the target releases the `Semaphore` permit.
2. At line 65, the target runs the blocking operation.
3. Concurrently, the fiber is cancelled. `trigger` runs, which sets `cb`, acquires the `Semaphore` permit, interrupts the target once, and releases the permit. It breaks out of the loop.
4. The canceller runs `repeat`, and reaches line 136. At this point, `done` is set to `false` and `cb` is not null, so the conditional fails and the carrier thread for the canceller exits the `async`.
5. The target thread receives the interrupt signal during its blocking operation, but (this is the key part!) it resets the interrupt status, so the interrupt signal is effectively suppressed! `cur.thunk()` returns normally.
6. At line 69, the target thread acquires the `Semaphore` permit successfully because the interrupt status isn't set, and so `InterruptedException` isn't thrown.
7. At line 84, the target thread reaches the `finally` block and sets `done` to true, but at this point it's too late.
8. `cb` was never invoked, so the canceller is perpetually blocked.

Whenever I'm pointing out these bugs, I should probably describe the interleaving of effects that produces a race condition. It makes it much easier for both of us to follow but also easier to point out if I'm wrong.

djspiewak replied:
> The canceller runs repeat, and reaches line 136. At this point, done is set to false and cb is not null, so the conditional fails and the carrier thread for the canceller exits the async.
>
> The target thread receives the interrupt signal during its blocking operation, but (this is the key part!) it resets the interrupt status, so the interrupt signal is effectively suppressed! cur.thunk() returns normally.

Ooooooooh this is a nice catch. I agree it's a problem. I wonder if we can solve it without introducing more state…

djspiewak replied:
See latest push. I think I fixed it. I realized that we don't have to *only* check the callback on `InterruptedException`

djspiewak committed:
Moved callback-running to happen regardless of cancelation

RaasAhsan reviewed (approved):
Looks good to me! What an adventure that was
